ocaml/database/db_lock.ml

```diff
@@ -11,60 +11,148 @@
+  type statistics = {
+      mutable max_time: float
+    ; mutable min_time: float
+    ; mutable total_time: float
+    ; mutable acquires: int
+  }
 
-let time = ref 0.0
+  val create : unit -> t
+  (** Creates an instance of a reentrant lock. *)
 
-let n = ref 0
+  val lock : t -> unit
+  (** [lock l] acquires the lock [l]. If the calling thread already
+      holds the lock, the implementation internally increases the number
+      of "holds" the thread has on the lock. Each call to [lock] must
+      have a corresponding call to [unlock] or else it is an error. *)
 
-let maxtime = ref neg_infinity
+  val unlock : t -> unit
+  (** [unlock l] releases a hold on the lock. If the hold count
+      becomes 0, the lock is free to be acquired by other threads. It is
+      an error to call this from a thread that does not hold the lock. *)
 
-let mintime = ref infinity
+  val statistics : t -> statistics
+  (** Returns a copy of the internal timing statistics maintained by
+      the implementation. Calling this has the effect of temporarily
+      acquiring the lock, as only the lock holder can read or modify the
+      internal record. *)
+end
 
-let thread_reenter_count = ref 0
+(** A simple re-entrant lock (recursive mutex). *)
+module ReentrantLock : REENTRANT_LOCK = struct
+  type tid = int
 
-let allow_thread_through_dbcache_mutex = ref None
+  type statistics = {
+      mutable max_time: float
+    ; mutable min_time: float
+    ; mutable total_time: float
+    ; mutable acquires: int
+  }
 
-let with_lock f =
-  let me = Thread.id (Thread.self ()) in
-  let do_with_lock () =
-    let now = Unix.gettimeofday () in
-    Mutex.lock dbcache_mutex ;
-    let now2 = Unix.gettimeofday () in
-    let delta = now2 -. now in
-    time := !time +. delta ;
-    n := !n + 1 ;
-    maxtime := max !maxtime delta ;
-    mintime := min !mintime delta ;
-    allow_thread_through_dbcache_mutex := Some me ;
-    thread_reenter_count := 1 ;
-    finally f (fun () ->
-        thread_reenter_count := !thread_reenter_count - 1 ;
-        if !thread_reenter_count = 0 then (
-          allow_thread_through_dbcache_mutex := None ;
-          Mutex.unlock dbcache_mutex
+  type t = {
+      holder: tid option Atomic.t (* The holder of the lock *)
+    ; mutable holds: int (* How many holds the holder has on the lock *)
+    ; lock: Mutex.t (* Barrier to signal waiting threads *)
+    ; condition: Condition.t
+          (* Waiting threads are signalled via this condition to reattempt to acquire the lock *)
+    ; statistics: statistics (* Bookkeeping of time taken to acquire lock *)
+  }
+
+  let create_statistics () =
+    {max_time= neg_infinity; min_time= infinity; total_time= 0.; acquires= 0}
+
+  let create () =
+    {
+      holder= Atomic.make None
+    ; holds= 0
+    ; lock= Mutex.create ()
+    ; condition= Condition.create ()
+    ; statistics= create_statistics ()
+    }
+
+  let current_tid () = Thread.(self () |> id)
+
+  let now = Unix.gettimeofday
```

Comment:
How important is backwards compatibility here? Could we drop the float and use the new time module instead?

Reply:
There;s quite a lot of code that should be converted from floats to mtime. I've done quite a bit in this dev branch: https://github.com/xapi-project/xen-api/compare/master...psafont:xen-api:private/paus/its-about-mtime

One such change can be found in this commit: https://github.com/xapi-project/xen-api/commit/18e91194c89e745be9be4f72971298cd826cf6be